Implement using web-streams-polyfill

MattiasBuelens opened the pull request:
This implements the WHATWG Streams API using the ES2018 ponyfill build from [web-streams-polyfill](https://github.com/MattiasBuelens/web-streams-polyfill).

The WPT set up is based off of @joyeecheung's work in #2. Many thanks for that! 😄

To do list:
* [x] The Node WPT runner starts running all the test scripts immediately. Since most of the streams tests are asynchronous (using `promise_test`), this means that tests can interfere with each other. This is especially bad for the "patched globals" tests, which will [literally break the API](https://github.com/web-platform-tests/wpt/blob/593d7eb21e1c3ac161aa51cce914056eb9182962/streams/readable-streams/patched-global.any.js#L78-L80) during the test.
    * I'm thinking we should make the runner execute tests *sequentially*, waiting for each test to complete before starting the next one? Concretely, this would mean we wait until [`completionCallback`](https://github.com/nodejs/node/blob/090ad7d4a1c32ed6fcd861ce215b1205afe4146c/test/common/wpt.js#L529) is called before we start the next test. I'll try implementing it here first, to see if it works. Hopefully we can upstream that afterwards? 🙏 
    * ✅ Fixed. The new test driver now runs each test in a separate worker, so they can no longer interfere with each other.
* [x] I need to copy a lot of global variables into the test runner. This is because the ponyfill gets loaded inside the original VM context, and uses globals such as `Promise` or `Uint8Array` from there. However, the tests are run inside a separate VM context, so when it tries to compare whether e.g. `ReadableStreamBYOBRequest.view` is a `Uint8Array`, it's using a *different* global for `Uint8Array` than the polyfill.
    * Is this the right approach? Or should we look for a way to e.g. load the ponyfill *inside* of the test VM context, so it uses that context's globals? It would be quite weird, though...
    * While most globals can be copied into the runner, some things just cannot be patched. For example, there's [a test](https://github.com/web-platform-tests/wpt/blob/593d7eb21e1c3ac161aa51cce914056eb9182962/streams/readable-streams/async-iterator.any.js#L23-L24) that checks if `ReadableStream[Symbol.asyncIterator]()` has [`%AsyncIteratorPrototype%`](https://tc39.es/ecma262/#sec-asynciteratorprototype) as its prototype. As far as I know, there's no way to replace that...
    * ✅ Fixed. The ponyfill is loaded into the worker's global scope at the start of each test, and uses the worker's global variables. No more copying global variables.
* [x] `ReadableStream.pipeTo()` uses an `AbortSignal`. Currently, I'm using [a separate polyfill](https://github.com/mysticatea/abort-controller) for that.
    * ~It looks like Node has recently added an experimental implementation for this in nodejs/node#33527? Could be interesting.~ 
    * ✅ Fixed. Now using the built-in `AbortSignal` from Node 15! 😁 
* [x] Similarly, `ReadableStream.pipeTo()` needs to reject with a `DOMException` when it is aborted using a signal, which is also tested by WPT.
    * While there is already [an implementation in Node](https://github.com/nodejs/node/blob/090ad7d4a1c32ed6fcd861ce215b1205afe4146c/lib/internal/per_context/domexception.js), it doesn't look like this is exposed globally. I suppose that if this streams implementation were to be added to Node itself, we could access it through an internal binding like with [the URL tests](https://github.com/nodejs/node/blob/090ad7d4a1c32ed6fcd861ce215b1205afe4146c/test/wpt/test-url.js#L8)?
    * ✅ Fixed. With `--expose-internals`, I can retrieve the `DOMException` implementation from Node itself. It's even given as [an example in the test driver's README](https://github.com/nodejs/node/blob/423490423264d1b7624d578a421a1bedecbc5d2e/test/wpt/README.md#3-create-the-test-driver)! 😅

MattiasBuelens commented:
Okay, the tests *really* don't like having the ponyfill come from a different VM. It breaks in all sorts of ways.

[`webidl2.js`](https://github.com/MattiasBuelens/whatwg-stream/blob/58c28a4f02fc040fdf066c689449bebcde81efac/test/fixtures/wpt/resources/webidl2/lib/webidl2.js#L1169) contains the following check:
```javascript
let proto = this;
while (proto !== Object.prototype) {
  // ...
  proto = Object.getPrototypeOf(proto);
}
```
However, the "parent VM's" `Object` is [injected](https://github.com/MattiasBuelens/whatwg-stream/blob/58c28a4f02fc040fdf066c689449bebcde81efac/test/common/wpt.js#L445) into the test VM, so any instance of a class that's defined inside the test VM (like [`Constructor`](https://github.com/MattiasBuelens/whatwg-stream/blob/58c28a4f02fc040fdf066c689449bebcde81efac/test/fixtures/wpt/resources/webidl2/lib/webidl2.js#L2502)) does not have *this* particular `Object.prototype` in its prototype chain... I could work around this by checking for `Object.getPrototypeOf(proto) !== null` instead, but it feels *weird*.

Then there's the [`exposed_in`](https://github.com/MattiasBuelens/whatwg-stream/blob/58c28a4f02fc040fdf066c689449bebcde81efac/test/fixtures/wpt/resources/idlharness.js#L784) function in `idlharness.js`, which checks what sort of environment we're running in to figure out which IDL definitions should be exposed. Of course, this function doesn't expect a Node-like environment, so it throws an error. Again, I could work around this by having it assume a worker-like environment, but it's yet another hack.

Finally, there are [these tests in `idlharness.js`](https://github.com/MattiasBuelens/whatwg-stream/blob/58c28a4f02fc040fdf066c689449bebcde81efac/test/fixtures/wpt/resources/idlharness.js#L1705-L1710) which check that an IDL object without a constructor cannot be called as a function or as a constructor. In that case, the VM throws an instance of *its original version* of `TypeError`, even if [we have already replaced the global `TypeError`](https://github.com/MattiasBuelens/whatwg-stream/blob/58c28a4f02fc040fdf066c689449bebcde81efac/test/wpt/test.js#L23). This causes a test failure on [this line](https://github.com/MattiasBuelens/whatwg-stream/blob/58c28a4f02fc040fdf066c689449bebcde81efac/test/fixtures/wpt/resources/testharness.js#L1576), since `e` is an instance of the test VM's original `TypeError`, while `constructor` is the global `TypeError` that has been replaced with the parent VM's version. I really don't know how to fix this one: if I let the test VM keep its own original `TypeError` instead, then all of the [other tests](https://github.com/MattiasBuelens/whatwg-stream/blob/58c28a4f02fc040fdf066c689449bebcde81efac/test/fixtures/wpt/streams/readable-streams/general.any.js#L23) that deal with `TypeError`s originating from the ponyfill (and thus coming from the parent VM) will fail... 😕

---

One way to fix these issues would be to load the ponyfill inside of the test VM, rather than loading it in the "parent VM" and copying it into the test VM's global scope. This would be closer to [what the reference implementation does with WPT runner](https://github.com/whatwg/streams/blob/0dbd3026df732a5bb8f66b23d9337244e5d0b3b9/reference-implementation/run-web-platform-tests.js#L56), but it would be very different from [the existing Node WPT tests](https://github.com/nodejs/node/blob/001a4a79284762102b8e89f3b254e2287be14a82/test/wpt/test-url.js#L12). I... don't really know how I feel about this. 🤷

targos commented:
Hi @MattiasBuelens !

I hit the same kind of issues as you when I tried to update the WPT test harness in https://github.com/nodejs/node/pull/33770.
I think I'm going to try and move the test execution to a child process instead of a VM instance.

MattiasBuelens commented:
@targos Interesting! So instead of spinning up a VM and copying variables into its global scope, the WPT runner would spin up a new Node process and have it load all of the necessary code (e.g. load the library and set up the global variables) to run the test?

targos commented:
Yes, exactly.

targos commented:
I went for a Worker-based approach, which was easier to implement. PR: https://github.com/nodejs/node/pull/34796

MattiasBuelens commented:
I see nodejs/node#34796 has landed recently. I'll have a look integrating the new test runner soon™, and see if that improves things. 🙂

MattiasBuelens committed:
Hack idlharness to use worker interfaces when running in our JS shell

MattiasBuelens commented on test/common/wpt.js:
```diff
@@ -0,0 +1,610 @@
+      if (key.includes('*')) {
+        this.patternMatch.push(new StatusRule(key, rules[key], key));
+      } else {
+        this.exactMatch[path.normalize(key)] = new StatusRule(key, rules[key]);
```
@targos I needed to add this `path.normalize()` call so it would correctly recognize the `"readable-byte-streams/bad-buffers-and-views.any.js"` rule on Windows. Should I upstream this to Node as well?

MattiasBuelens replied:
This was fixed in nodejs/node#37467. 😁

MattiasBuelens commented on test/fixtures/wpt/resources/idlharness.js:
```diff
@@ -0,0 +1,3599 @@
+        // HACK: Pretend that our JavaScript shell exposes worker interfaces.
+        // The streams interfaces are currently only exposed for windows, workers and worklets
+        // and there's no such thing as [Exposed=Shell] in Web IDL yet.
+        return globals.has("DedicatedWorker");
```
Better suggestions are welcome. 😁

MattiasBuelens reviewed (commented):
Progress! With the new WPT test runner (and only a *few* hacks), I can get all relevant tests to pass. 🎉

The polyfill still needs to be updated to handle [one of the more recent spec changes](https://github.com/whatwg/streams/commit/6cd5e81f6191fed9e7d99ee73d4941e3060311ce). After that, I'll run `git node wpt streams` again to get the latest WPT streams tests as well.

Ethan-Arrowood commented:
Hi @MattiasBuelens can I help you get this across the finish line?

MattiasBuelens commented on test/wpt/status/streams.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "piping/pipe-through.any.js": {
+    "fail": "Node does not perform a brand check for AbortSignal.prototype.aborted"
```
[This test](https://github.com/web-platform-tests/wpt/blob/887350c2f46def5b01c4dd1f8d2eee35dfb9c5bb/streams/piping/pipe-through.any.js#L176) checks whether passing a "bad" `AbortSignal` (in particular ) causes `pipeThrough()` to throw. [The polyfill implements this](https://github.com/MattiasBuelens/web-streams-polyfill/blob/6915a01aa6c92b4d3cb306fb7a66299e881a7bcc/src/lib/abort-signal.ts#L33-L38) by trying to retrieve `AbortSignal.prototype.aborted` and checking that it does not throw. [WebIDL requires this attribute to perform a brand check](https://heycam.github.io/webidl/#dfn-attribute-getter), so a compliant implementation *should* throw a `TypeError` when this getter is run on such a "bad" `AbortSignal`.

Unfortunately, Node's [AbortSignal implementation](https://github.com/nodejs/node/blob/d15475578abfcc7e0a420719b8d55a6b3ab4f198/lib/internal/abort_controller.js#L46) **does not** throw an error when `AbortSignal.prototype.aborted` is run such a signal. Instead, it will return `false` and the polyfill will consider this to be a valid `AbortSignal`.

I'm not sure what we want to do here. Should we fix Node to perform a brand check in `get aborted` (i.e. throw an error if `this[kAborted]` does not exist)? Or should we change how the polyfill does this check (although I'm not sure how)?

Ethan-Arrowood replied:
If I understand you correctly, I think the best course of action is to make an upstream change to Node.js core AbortSignal so that it aligns more with the spec.

MattiasBuelens replied:
On it! 🙂

**EDIT:** Opened nodejs/node#37720.

MattiasBuelens reviewed (commented):
@Ethan-Arrowood Sorry, this has been lingering for a while. I didn't get much feedback, so I thought people weren't very interested. 🤷‍♂️

Either way, I've just updated the PR to the latest version of the polyfill and the WPT tests. Have a look!

Ethan-Arrowood commented:
Thank you for pushing all of these updates! I'll review next week when I have some more time to go through it all. I'd also like to encourage @nodejs/whatwg-stream and @nodejs/undici to take a look here. Now that things likes EventTarget and AbortController are in core this is the next logical progression towards Fetch and overall WHATWG Stream support for Node 😄

mcollina commented:
Given that this is based on http://npm.im/web-streams-polyfill, is having this repo still useful?

Are you running the WPT tests in   http://npm.im/web-streams-polyfill?

aduh95 commented on test/wpt/test.js:
```diff
@@ -0,0 +1,33 @@
+// Note:
+// - The --expose-internals flag is specific to Node.js core's
+//   internal error system and is unnecessary for external users.
+// - The --experimental-worker flag is used to inform the
+//   WPT harness that we are running tests in the Worker scope
+//   so it doesn't use document.methods().
```
Maybe it's only useful in some environments? On my machine I tried to remove the flag and got the same result with or without it.

MattiasBuelens replied:
You're right. Workers are now enabled by default in Node, so we no longer need this flag. I'll remove it. 🙂

ronag commented:
This is a bit hard to review with so many files.

MattiasBuelens commented:
@ronag Note that everything inside `test/fixtures/wpt/` is copied directly from the [web platform tests](https://github.com/web-platform-tests/wpt) by `git node wpt`, and should not need review. See the explainer in [test/fixtures/wpt/README.md](https://github.com/nodejs/whatwg-stream/blob/51035a994bff8e6ebddd0de72da11523623e6010/test/fixtures/wpt/README.md).

MattiasBuelens commented:
@mcollina Yes, [the polyfill also runs the WPT tests](https://github.com/MattiasBuelens/web-streams-polyfill/blob/v3.0.2/test/run-web-platform-tests.js).

The polyfill uses a different test runner though: `wpt-runner` (based on `jsdom`) instead of [Node's own `WPTRunner`](https://github.com/nodejs/node/blob/v15.11.0/test/wpt/README.md). So this PR demonstrates that we *can* integrate with Node's runner as well.

ronag commented:
> @ronag Note that everything inside `test/fixtures/wpt/` is copied directly from the [web platform tests](https://github.com/web-platform-tests/wpt) by `git node wpt`, and should not need review. See the explainer in [test/fixtures/wpt/README.md](https://github.com/nodejs/whatwg-stream/blob/51035a994bff8e6ebddd0de72da11523623e6010/test/fixtures/wpt/README.md).

Yea, I get that but it's still difficult to review with the GitHub UI. Could you perhaps put all the functionality changes in one commit which we can review and the test fixtures in another one?

MattiasBuelens committed:
Implement using web-streams-polyfill

MattiasBuelens committed:
Add WPT fixtures

git node wpt streams
git node wpt resources
git node wpt interfaces
git node wpt common

MattiasBuelens committed:
Add WPT runner

MattiasBuelens committed:
Set up tests

MattiasBuelens commented:
@ronag I've squashed the history together in a few commits. All WPT fixtures are now added in one single commit, so you can skip that one.

mcollina reviewed (approved):
lgtm

mcollina commented:
@joyeecheung @targos @benjamingr should we land this?

benjamingr commented:
I'm in favour

targos commented:
What's going to happen after we land? I don't remember the goal of this repository, sorry 😄

mcollina commented:
provide a stopgap solution while we add them to core and bless something as our official implementation. I'll open an issue on nodejs/node.

targos commented:
But if I understand correctly, this delegates the implementation to an existing module, so the stopgap solution already exists? I'm okay with that as a temporary thing but in the end we will not be able to integrate it like this in Node.js core (the code will have to be copied/adapted at some point).

mcollina commented:
> But if I understand correctly, this delegates the implementation to an existing module, so the stopgap solution already exists? I'm okay with that as a temporary thing but in the end we will not be able to integrate it like this in Node.js core (the code will have to be copied/adapted at some point).

Yes exactly.

MattiasBuelens commented:
Now that Node has a built-in experimental implementation of web streams with nodejs/node#39062 (🎉🎉🎉), I believe this PR and (by extension) this repository has become obsolete.

I suggest we close this PR and/or archive this repository. No hard feelings, on the contrary: I'm very happy that web streams made it into Node *proper*! 😄

jimmywarting commented:
Woho good news!
